reference/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,300 @@
+
+    // Only those strategies are supported here yet, assert this.
+    GKO_ASSERT(strategy == strategies::minimum_degree ||
+               strategy == strategies::pseudo_peripheral);
+
+    // There must always be at least one unvisited node left.
+    auto index_min_node = 0;
+    auto min_node_degree = std::numeric_limits<IndexType>::max();
+    for (auto i = 0; i < num_vertices; ++i) {
+        if (!visited[i]) {
+            if (degrees[i] < min_node_degree) {
+                index_min_node = i;
+                min_node_degree = degrees[i];
+            }
+        }
+    }
+
+    // If that is all that is required, return.
+    if (strategy == gko::reorder::starting_strategy::minimum_degree) {
+        return index_min_node;
+    }
+
+    // Find a pseudo-peripheral node, starting from an node of minimum degree.
+    auto current = index_min_node;
+
+    // Isolated nodes are by definition peripheral.
+    if (min_node_degree == 0) {
+        return index_min_node;
+    }
+
+    // This algorithm is e. g. presented in
+    // http://heath.cs.illinois.edu/courses/cs598mh/george_liu.pdf, p. 70
+    // onwards.
+    auto contender_and_height = rls_contender_and_height<IndexType>(
+        exec, num_vertices, current, row_ptrs, col_idxs, degrees);
+    while (true) {
+        auto contender_and_height_contender =
+            rls_contender_and_height<IndexType>(exec, num_vertices,
+                                                contender_and_height.first,
+                                                row_ptrs, col_idxs, degrees);
+
+        if (contender_and_height_contender.second >
+            contender_and_height.second) {
+            contender_and_height.second = contender_and_height_contender.second;
+            current = contender_and_height.first;
+            contender_and_height.first = contender_and_height_contender.first;
+        } else {
+            return current;
+        }
+    }
+}
+
+/**
+ * Comutes a rcm reording using a naive sequential algorithm.
+ */
+template <typename IndexType>
+void get_permutation(std::shared_ptr<const ReferenceExecutor> exec,
+                     const size_type num_vertices,
+                     const IndexType *const row_ptrs,
+                     const IndexType *const col_idxs,
+                     const IndexType *const degrees,
+                     IndexType *const permutation,
+                     IndexType *const inv_permutation,
+                     const gko::reorder::starting_strategy strategy)
+{
+    // Storing vertices left to proceess.
+    Array<IndexType> linear_queue(exec, num_vertices);
+    auto linear_queue_p = linear_queue.get_data();
+    auto head_offset = 0;
+    auto tail_offset = 0;
+
+    // Storing which vertices have already been visited.
+    vector<bool> visited(num_vertices, false, exec);
+
+    for (auto perm_index = 0; perm_index < num_vertices; ++perm_index) {
+        // Get the next vertex to process.
+        auto next_vertex = 0;
```

Comment:
I assume this is supposed to `size_type`:
```suggestion
        size_type next_vertex = 0;
```